src/Features/Core/Portable/Testing/AbstractTestMethodFinder.cs

```diff
@@ -0,0 +1,107 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Features.Testing;
+
+internal abstract class AbstractTestMethodFinder<TMethodDeclaration>(IEnumerable<ITestFrameworkMetadata> testFrameworks) : ITestMethodFinder where TMethodDeclaration : SyntaxNode
+{
+    /// <summary>
+    /// Output the method symbol as a fully qualified method name, e.g. Namespace.Class.Method to match what test discovery gives us.
+    /// Generics are not applicable here - none of our supported test frameworks allow generic test classes / methods.
+    /// </summary>
+    private static readonly SymbolDisplayFormat s_methodSymbolNoParametersDisplayFormat = new(
+        typeQualificationStyle: SymbolDisplayTypeQualificationStyle.NameAndContainingTypesAndNamespaces,
+        memberOptions: SymbolDisplayMemberOptions.IncludeContainingType);
+
+    protected readonly ImmutableArray<ITestFrameworkMetadata> TestFrameworkMetadata = testFrameworks.ToImmutableArray();
+
+    protected abstract bool IsTestMethod(TMethodDeclaration method);
+
+    protected abstract bool DescendIntoChildren(SyntaxNode node);
+
+    public async Task<ImmutableArray<SyntaxNode>> GetPotentialTestMethodsAsync(Document document, TextSpan textSpan, CancellationToken cancellationToken)
+    {
+        var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+        var testNodes = await GetPotentialTestNodesAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+
+        // Find any test methods that intersect with the requested span.
+        var intersectingNodes = testNodes.Where(node => node.Span.IntersectsWith(textSpan)).ToImmutableArray();
+        if (!intersectingNodes.IsEmpty)
+        {
+            return intersectingNodes;
+        }
+
+        // We might have been invoked on a test class.  Check if any of the test method parent nodes intersect with the requested text span.
+        return testNodes.Where(node => node.Parent?.Span.IntersectsWith(textSpan) == true).ToImmutableArray();
+    }
+
+    public bool IsMatch(SemanticModel semanticModel, SyntaxNode node, string fullyQualifiedTestName, CancellationToken cancellationToken)
+    {
+        var method = node as TMethodDeclaration;
+        Contract.ThrowIfNull(method, $"Node with kind {node.RawKind} is not a method");
+
+        // Since discovered tests are not guarantied to run on a particular snapshot, we match optimistically based on test name.
+        var methodSymbol = semanticModel.GetDeclaredSymbol(method, cancellationToken);
+        Contract.ThrowIfNull(methodSymbol, "Test method has no symbol");
+
+        var fullyQualifiedMethodName = methodSymbol.ToDisplayString(s_methodSymbolNoParametersDisplayFormat);
+
+        // The definition of fully qualified name varies depending on the test framework.
```

Comment:
https://github.com/microsoft/vstest/blob/main/docs/RFCs/0017-Managed-TestCase-Properties.md

Will this logic be improved by using this?
I think VSTest provided a standard fully qualified name to managed test cases

Reply:
YES!  That looks great, I will try it out.  Thanks for linking that

Reply:
Unfortunately those properties aren't public, and even if they were they appear to be unset.  So can't use them :(
![image](https://github.com/dotnet/roslyn/assets/5749229/2d718fd3-82ba-4e13-b89a-f23cbbe46e60)

Reply:
Emm I think there should be some way to use that.
@shyamnamboodiripad Do you know how to ask TP to generate these names? Or is the FQN work still not fully implemented yet?

Reply:
@Cosifne @dibarbet That would depend on support for the FQN properties being implemented in all test adapters. As of now, I believe only the latest versions of MSTest adapter support this. xUnit and NUnit don't support this yet. cc @peterwald